src/base/static/libs/maps/mapboxgl-provider.js

```diff
@@ -173,29 +257,98 @@ export default (container, options) => {
       });
     },
 
-    addGeoJSONLayer: (layerStyles, geometryType) => {
-      // TODO: Can we detect GeoJSON geometry types automatically instead?
-      // What about datasets with multiple geometry types?
-      // It's possible to obtain the geometry type in filter expressions, e.g.:
-      // ["==", "$type", "Point"]
-
-      if (geometryType === "Point") {
-        layerStyles.forEach(layerStyle => {
-          layerStyle.layout["icon-allow-overlap"] = true;
-          layerStyle.type = "symbol";
+    addGeoJSONLayer: layerStyles => {
```

Comment:
To add some context here because I think this part of the code might be confusing-- 

The idea here is to handle two situations: GeoJSON objects made up of combinations of `Point`, `Polygon`, and `LineString` geometry, and layer style information that's stored along with a place in the database.

Both of these situations exist because we support the geometry editor. Users can build any kind of geometry and style it independently of style rules defined in the config. We store the styling information in the model JSON blob under a property called `style`. If that property exists, we use it to set `paint` and `layout` Mapbox properties.

One TODO here is that custom icon urls are stored as relative file paths in the database, e.g.:
```
"style": {
  "iconUrl": "/static/css/images/markers/marker-star.png"
}
```

This is brittle, but I think Mapbox GL presents a nice solution: store the sprite identifier for a custom icon instead of a file path. So this PR makes that assumption, but I'll need to refactor icon file paths on the database in coordination with the deployment of the `Map Port` feature branch.

Reply:
> Why are we adding 3 layers for every layerStyle that we pass in?

We add one layer for each geometry type that might be in the geojson: `marker` for `Point` geometry, `line` for `LineString`, and `fill` for Polygon. 

I think an alternative approach would be to filter the geojson before we create a Mapbox source from it, such that we have a separate source for all `Point`s, `LineString`s, and `Polygons`. Then we can make a separate layer for each without the need for the `geometry-type` filters. I'm not sure which is better, though the Mapbox filters seem very powerful so it might make sense to lean on them.

> For the custom urls, it looks like we will have to migrate all references of icon_url and iconUrl in our config, as well as in our db

Yes-- I think the final map PR will be a giant config refactor for flavors we want to support. As for the database refactoring, my guess is that doing it manually will be easier in this case. There aren't a lot of places with custom urls.

Reply:
>We add one layer for each geometry type that might be in the geojson: marker for Point geometry, line for LineString, and fill for Polygon.

But it seems like we should be able to infer which Geometry Type belongs to each Layer Style by examining the Layer Style. For example:

```js
layerStyles
  .filter(layerStyle => layerStyle.layout)
  .map(layerStyle => {
    return {
      id: `${layerStyle.id}_symbol`,
      source: layerStyle.source,
      // etc
```

where I just added a `.filter(layerStyle => layerStyle.layout)` filter step.  

And for our Line geometries:

```javascript
layerStyles
  .filter(
    layerStyle =>
      layerStyle.paint &&
      (layerStyles.paint["line-color"] || layerStyles.paint["line-opacity"]),
  )
  .map(layerStyle => {
    return {
      id: `${layerStyle.id}_line`,
      source: layerStyle.source,
      // etc.
```

Does that make sense? Doing so should allow us to remove the `layerStyle.paint &&` and `layerStyle.layout &&` checks as well as having to add a `filter`.

But maybe I am missing something. Perhaps these `filter` steps would not work for LayerStyles generated by the Geometry Editor? What do you think?

Reply:
Well, I think you still need a Mapbox `filter` expression for the `geometry-type`, since all these layers refer to the same geojson source object and that source might be made up of any combination of `Point`, `LineString`, or `Polygon` geometry.

It seems like there should be a way to simplify all the `layout` and `paint` blocks. I don't think the approach above will work, though, because somewhere you still need to do this check (e.g. for `Point` geometry):

```
"icon-image": (layerStyle.layout &&
  layerStyle.layout["icon-image"]) || [
  "get",
  "iconUrl",
  ["get", "style"],
],
```
which looks to see if there is an `icon-image` set in the config (`layerStyle.layout && layerStyle.layout["icon-image"]`) and if not falls back to the `iconUrl` found in the `style` property (which originates from the database).

Still, it's messy, and it also doesn't allow us to set `paint` and `layout` properties that aren't explicitly referenced here. (For example, what if we want to set something like the [`line-pattern`](https://www.mapbox.com/mapbox-gl-js/style-spec/#paint-line-line-pattern) paint property for line geometry? Or anything else?)

I'll think about this some more--there must be a better pattern...

Reply:
Ah, thanks for the explanation - you are right about still needing the `filter` expression for the `geometry-type`. I do think there is a better pattern, but it might not be worth exploring until we finish the map port, including the geometry editor port.

I think once we have a better understanding about what properties we can expect in our `layerStyle`'s, then we can clean this up. But for now, it might not be a priority.

Reply:
Maybe we can do something like this:

```
layerStyles
      .map(layerStyle => {
        const fallbackIconImage =
          (layerStyle.layout && layerStyle.layout["icon-image"]) ||
          "no-icon-image";
        return {
          id: `${layerStyle.id}_symbol`,
          source: layerStyle.source,
          layout: {
            "icon-image": [
              "case",
              ["all", ["has", "style"], ["has", "iconUrl", ["get", "style"]]],
              ["get", "iconUrl", ["get", "style"]],
              fallbackIconImage,
            ],
            "icon-allow-overlap": true,
            ...layerStyle.layout,
          },
          type: "symbol",
          filter: _appendFilters(layerStyle.filter, [
            "==",
            ["geometry-type"],
            "Point",
          ]),
        };
      })
```

So the intended behavior (for `Point` geometry) is: 
* if a feature has an `iconUrl` stored in the database, use that
* if not, fall back to a config-defined icon image
* enforce the `icon-allow-overlap: true` `layout` property
* pass through any other `layout` properties that are set in the config

We would follow a similar pattern for `layout` and `paint` properties for both `LineString` and `Polygon` geometry.

Do you think that would work?

Reply:
I'm not sure that I completely follow the differences between the `  ["all", ["has", "style"], ["has", "iconUrl", ["get", "style"]]],` and `["get", "iconUrl", ["get", "style"]]` expressions, but this definitely seems like an improvement.

I think a lot of the complexity comes from us not having a proper Layer abstraction, and this part of the code should not be concerned with what is stored in the Place vs the config. But we can address that issue later.

Reply:
`["all", ["has", "style"], ["has", "iconUrl", ["get", "style"]]]` resolves to a boolean, which is needed by the `case` operator. I think it's basically the same as `!!(layerStyle.style && layerStyle.iconUrl)`.

> I think a lot of the complexity comes from us not having a proper Layer abstraction

Agreed. I might try to make some progress in that direction in the next PR.